File: PitBoard(Android)/app/src/main/java/com/pitboard/app/standings/sources/ImsaStandingsSource.kt
package com.pitboard.app.standings.sources

import com.pitboard.app.standings.CarDriverEntity
import com.pitboard.app.standings.StandingEntity
import com.pitboard.app.standings.StandingType
import com.pitboard.app.standings.StandingsCategory
import com.pitboard.app.standings.StandingsClass
import com.pitboard.app.standings.StandingsHttpClient
import com.pitboard.app.standings.StandingsMoshi
import com.squareup.moshi.Types
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import okhttp3.FormBody
import okhttp3.Request
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import java.time.Year

/**
 * IMSA WeatherTech SportsCar Championship: igual que ELMS, se trata "por coche" dentro
 * de sus 4 clases (GTP, LMP2, GTD Pro, GTD — LMP3 salió del campeonato principal para
 * 2026). A diferencia de ELMS, no hay una única página con todo: comprobado a mano el
 * 02/09/2026 contra la web real, con curl (no es una suposición).
 *
 * NO implementa StandingsSource: esa interfaz devuelve solo `List<StandingEntity>`, pero
 * aquí el logo de equipo (que va EN esas filas) sale de la misma página que los pilotos
 * (ver más abajo), así que separar "clasificación" y "pilotos" en dos fuentes
 * independientes habría obligado a repetir la visita a las ~50 páginas de equipo dos
 * veces, o a que una fuente dependiera del resultado de la otra a medio terminar. Con
 * todo en una sola clase, cada equipo se visita una vez y las dos salidas
 * (`standings` + `carDrivers`) salen ya completas. StandingsRepository la trata como una
 * rama de sincronización aparte, con el mismo aislamiento de fallos que las demás.
 *
 * 1) CLASIFICACIÓN POR COCHE — imsa.com usa un widget AJAX de WordPress:
 *    `POST https://www.imsa.com/wp-admin/admin-ajax.php` con
 *    `action=getImsaStandings&standings=team&seriesId=1&currentYear={año}&classId={id}`.
 *    La respuesta es una tabla HTML con `#NN Team Name` por fila (número de coche +
 *    equipo) y el total de puntos en `td.totalpoints`. El `classId` de cada clase no
 *    está hardcodeado: se resuelve por su `shortcode` ("GTP", "LMP2"...) contra
 *    `https://dvw6yynr86g3k.cloudfront.net/galaxy/api/classes?series_id=1&season_id={año}`,
 *    una API JSON pública y estable — así que si IMSA renumerase los ids de clase de una
 *    temporada a otra, esto se sigue resolviendo solo. Solo si esa API fallara se cae a
 *    los ids comprobados hoy (FALLBACK_CLASS_IDS).
 *
 * 2) PILOTOS + LOGO POR COCHE — cada fila anterior enlaza a
 *    `imsa.com/racing-teams/{slug}/`, y esa página trae:
 *      - `.team-logos img`: 2-3 logos (serie/equipo/fabricante). CORREGIDO 02/09/2026:
 *        el primer intento asumía que el logo de equipo era el único cuya URL contiene
 *        "TeamLogo" — cierto para "13 Autosport" (2026_13Autosport_TeamLogo_282x188.png)
 *        pero NO para el resto: comprobado en 6 equipos reales (AF Corse, BMW Team RLL,
 *        Aston Martin THOR, Sun Energy 1, Meyer Shank, 13 Autosport) que el nombre de
 *        archivo del logo de equipo varía sin ningún patrón común. Lo que SÍ es
 *        constante en los 6: la 1ª imagen es siempre el logo fijo de la serie
 *        ("weathertech_championship.png") y la 2ª es siempre el logo del equipo (a
 *        veces "nologo_0.jpg" — el propio placeholder de IMSA para un equipo sin logo
 *        real, que aquí se trata como "sin logo" igual que el resto de la app). Ahora se
 *        descartan esas dos por nombre (serie fija / placeholder) y se coge la primera
 *        que quede, en vez de buscar "TeamLogo" en la URL.
 *      - Una tarjeta `div.imsa-card_item_widget` por piloto, con `img[data-src]` (la
 *        foto real — el `src` es un placeholder gris de carga perezosa) y
 *        `p.imsa-ciw-title` (nombre).
 *    Se descartó agrupar pilotos por el `man-id-NNNN` que trae la tabla de PILOTOS de
 *    imsa.com/weathertech/standings/: comprobado que ese id es del FABRICANTE (ej.
 *    Porsche), no del coche — un fabricante con 2 coches en la misma clase agruparía
 *    mal a sus pilotos. Por eso todo sale de la tabla de EQUIPOS + su página, nunca de
 *    esa tabla de pilotos.
 *
 * HONESTO sobre el coste: a diferencia de ELMS (una sola página, 140 pilotos), aquí hace
 * falta una petición por coche — unos 45-50 en total. Para no golpear el servidor de
 * golpe se lanzan en bloques de 8 en paralelo, nunca las ~50 a la vez. Un fallo al
 * visitar la página de un coche concreto solo deja a ESE coche sin logo/pilotos (sigue
 * apareciendo en la clasificación, con el icono por defecto) — nunca tumba la
 * sincronización completa.
 */
class ImsaStandingsSource {

    val category = StandingsCategory.IMSA

    suspend fun fetch(nowUtc: Long): ImsaFetchResult {
        val year = Year.now().value
        val classIds = resolveClassIds(year)

        val teamRows = CLASS_SHORTCODES.entries.flatMap { (standingsClass, shortcode) ->
            val classId = classIds[shortcode] ?: FALLBACK_CLASS_IDS.getValue(standingsClass)
            runCatching { fetchClassTeamRows(standingsClass, classId, year) }.getOrElse { emptyList() }
        }

        val teamPages = coroutineScope {
            teamRows.chunked(TEAM_PAGE_CONCURRENCY).flatMap { chunk ->
                chunk.map { row -> async(Dispatchers.IO) { row to row.teamUrl?.let { fetchTeamPage(it, row, nowUtc) } } }
                    .awaitAll()
            }
        }.toMap()

        val standings = teamRows.map { row ->
            val logoUrl = teamPages[row]?.logoUrl
            StandingEntity(
                category = category,
                standingsClass = row.standingsClass,
                type = StandingType.TEAM,
                entrantKey = "${category.name}-${row.standingsClass.name}-TEAM-${row.teamName}-${row.carNumber}",
                position = row.position,
                name = "#${row.carNumber}",
                team = row.teamName,
                points = row.points,
                photoUrl = logoUrl,
                updatedAtUtc = nowUtc
            )
        }

        val carDrivers = teamPages.values.filterNotNull().flatMap { it.drivers }

        return ImsaFetchResult(standings, carDrivers)
    }

    /** Resuelve el classId de cada shortcode ("GTP", "LMP2"...) contra la API pública de
     *  clases — mapa vacío (nunca excepción) si esa API fallara, para caer al fallback. */
    private fun resolveClassIds(year: Int): Map<String, String> = runCatching {
        val url = "https://dvw6yynr86g3k.cloudfront.net/galaxy/api/classes?series_id=$SERIES_ID&season_id=$year"
        val request = Request.Builder().url(url).header("User-Agent", USER_AGENT).build()
        val json = StandingsHttpClient.instance.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("$url: HTTP ${response.code}")
            response.body?.string() ?: error("$url: cuerpo vacío")
        }

        val listType = Types.newParameterizedType(List::class.java, ImsaGalaxyClass::class.java)
        val adapter = StandingsMoshi.instance.adapter<List<ImsaGalaxyClass>>(listType)
        adapter.fromJson(json).orEmpty()
            .filter { it.shortcode != null }
            .associate { it.shortcode!!.uppercase() to it.id.toString() }
    }.getOrElse { emptyMap() }

    private fun fetchClassTeamRows(standingsClass: StandingsClass, classId: String, year: Int): List<TeamRow> {
        val body = FormBody.Builder()
            .add("action", "getImsaStandings")
            .add("standings", "team")
            .add("seriesId", SERIES_ID)
            .add("currentYear", year.toString())
            .add("classId", classId)
            .build()

        val request = Request.Builder()
            .url(AJAX_URL)
            .header("User-Agent", USER_AGENT)
            .header("Referer", "https://www.imsa.com/weathertech/standings/")
            .post(body)
            .build()

        val html = StandingsHttpClient.instance.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("imsa standings ($standingsClass): HTTP ${response.code}")
            response.body?.string() ?: error("imsa standings ($standingsClass): cuerpo vacío")
        }

        val doc = Jsoup.parse(html)
        // Solo la tabla "desktop" trae puntos (td.totalpoints) — la "mobile" repite las
        // mismas filas sin esa columna, así que acotar por esa celda evita procesar cada
        // coche dos veces.
        return doc.select("tr:has(td.totalpoints)").mapIndexedNotNull { index, row ->
            parseTeamRow(row, standingsClass, index + 1)
        }
    }

    /** "#04 Crowdstrike Racing by APR" -> número de coche + nombre de equipo. Algunas
     *  filas no traen el `<a>` de enlace (equipo sin ficha propia todavía) — se lee igual
     *  el texto de la celda, solo que esa fila se queda sin logo ni pilotos. */
    private fun parseTeamRow(row: Element, standingsClass: StandingsClass, position: Int): TeamRow? {
        val cell = row.selectFirst("td.team-col") ?: return null
        val text = cell.text().trim()
        val match = CAR_NUMBER_TEAM.find(text) ?: return null
        val carNumber = match.groupValues[1]
        val teamName = match.groupValues[2].trim()
        if (teamName.isBlank()) return null

        val points = row.selectFirst("td.totalpoints")?.text()?.trim()?.toDoubleOrNull() ?: 0.0
        val teamUrl = cell.selectFirst("a.team-name")?.absUrl("href")?.takeIf { it.isNotBlank() }

        return TeamRow(standingsClass, position, carNumber, teamName, points, teamUrl)
    }

    private fun fetchTeamPage(teamUrl: String, row: TeamRow, nowUtc: Long): TeamPage? = runCatching {
        // El Referer evita algún 403 puntual comprobado en pruebas manuales (parece
        // bastarle con que la petición diga venir de la propia imsa.com).
        val request = Request.Builder()
            .url(teamUrl)
            .header("User-Agent", USER_AGENT)
            .header("Referer", "https://www.imsa.com/weathertech/teams/")
            .build()
        val html = StandingsHttpClient.instance.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("$teamUrl: HTTP ${response.code}")
            response.body?.string() ?: error("$teamUrl: cuerpo vacío")
        }
        val doc = Jsoup.parse(html, teamUrl)

        val logoUrl = doc.select(".team-logos img").map { it.absUrl("src") }
            .firstOrNull {
                !it.contains("weathertech_championship", ignoreCase = true) &&
                    !it.contains("nologo", ignoreCase = true)
            }

        val drivers = doc.select("div.imsa-card_item_widget").mapNotNull { card ->
            val name = card.selectFirst("p.imsa-ciw-title")?.text()?.trim().orEmpty()
            if (name.isBlank()) return@mapNotNull null
            val photoUrl = card.selectFirst("img.imsa-ciw-image")?.absUrl("data-src")?.takeIf { it.isNotBlank() }

            CarDriverEntity(
                category = StandingsCategory.IMSA,
                standingsClass = row.standingsClass,
                carNumber = row.carNumber,
                entryKey = normalize(name),
                name = name,
                photoUrl = photoUrl,
                updatedAtUtc = nowUtc
            )
        }

        TeamPage(logoUrl, drivers)
    }.getOrNull()

    private fun normalize(s: String): String =
        java.text.Normalizer.normalize(s, java.text.Normalizer.Form.NFD)
            .replace(Regex("\\p{Mn}+"), "")
            .lowercase()
            .replace(Regex("[^a-z0-9\\s]"), "")
            .replace(Regex("\\s+"), " ")
            .trim()

    private data class TeamRow(
        val standingsClass: StandingsClass,
        val position: Int,
        val carNumber: String,
        val teamName: String,
        val points: Double,
        val teamUrl: String?
    )

    private data class TeamPage(val logoUrl: String?, val drivers: List<CarDriverEntity>)

    companion object {
        private const val USER_AGENT = "Mozilla/5.0 (Linux; Android 14) PitBoard/1.0"
        private const val AJAX_URL = "https://www.imsa.com/wp-admin/admin-ajax.php"
        /** WeatherTech SportsCar Championship — comprobado en galaxy/api/series (id 1;
         *  el 2 y el 3 son las otras dos series de IMSA, IMPC y VPRC). */
        private const val SERIES_ID = "1"
        /** No más de 8 páginas de equipo a la vez (de las ~45-50 en total). */
        private const val TEAM_PAGE_CONCURRENCY = 8

        private val CAR_NUMBER_TEAM = Regex("^#(\\d+)\\s+(.+)$")

        private val CLASS_SHORTCODES: Map<StandingsClass, String> = mapOf(
            StandingsClass.GTP to "GTP",
            StandingsClass.LMP2 to "LMP2",
            StandingsClass.GTD_PRO to "GTD PRO",
            StandingsClass.GTD to "GTD"
        )

        /** Ids comprobados a mano el 02/09/2026 para la temporada 2026 — solo se usan si
         *  la API de clases (ver resolveClassIds) no responde. */
        private val FALLBACK_CLASS_IDS: Map<StandingsClass, String> = mapOf(
            StandingsClass.GTP to "194",
            StandingsClass.LMP2 to "196",
            StandingsClass.GTD_PRO to "192",
            StandingsClass.GTD to "191"
        )
    }
}

data class ImsaFetchResult(
    val standings: List<StandingEntity>,
    val carDrivers: List<CarDriverEntity>
)

data class ImsaGalaxyClass(val id: Int, val shortcode: String?)

File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Standings/Sources/ImsaStandingsSource.swift
import Foundation
import SwiftSoup

/// IMSA WeatherTech SportsCar Championship — equivalente exacto de
/// `ImsaStandingsSource.kt`. Igual que ELMS, se trata "por coche" dentro de sus 4 clases
/// (GTP, LMP2, GTD Pro, GTD).
///
/// No conforma `StandingsSource` (misma razón que en Android): el logo de equipo va EN
/// las filas de clasificación, pero sale de la MISMA página que los pilotos — separar
/// "clasificación" y "pilotos" en dos fuentes obligaría a visitar las ~50 páginas de
/// equipo dos veces. `StandingsRepository` la trata como una rama de sincronización
/// aparte, con el mismo aislamiento de fallos que las demás.
///
/// 1) CLASIFICACIÓN POR COCHE — widget AJAX de WordPress:
///    `POST https://www.imsa.com/wp-admin/admin-ajax.php` con
///    `action=getImsaStandings&standings=team&seriesId=1&currentYear={año}&classId={id}`.
///    El `classId` de cada clase se resuelve por su `shortcode` contra una API JSON
///    pública y estable; si esa API fallara se cae a `fallbackClassIds`.
/// 2) PILOTOS + LOGO POR COCHE — cada fila enlaza a `imsa.com/racing-teams/{slug}/`, con
///    `.team-logos img` (se descartan el logo fijo de la serie y el placeholder
///    "nologo", se queda con la siguiente) y tarjetas `div.imsa-card_item_widget` con
///    `img[data-src]` (foto real — el `src` es un placeholder de carga perezosa) y
///    `p.imsa-ciw-title` (nombre).
public final class ImsaStandingsSource: @unchecked Sendable {

    public let category: StandingsCategory = .imsa

    public init() {}

    public func fetch(nowUtc: Date) async throws -> ImsaFetchResult {
        let year = Calendar(identifier: .gregorian).component(.year, from: Date())
        let classIds = await resolveClassIds(year: year)

        var teamRows: [TeamRow] = []
        for (standingsClass, shortcode) in Self.classShortcodes {
            let classId = classIds[shortcode] ?? Self.fallbackClassIds[standingsClass] ?? ""
            guard !classId.isEmpty else { continue }
            if let rows = try? await fetchClassTeamRows(standingsClass: standingsClass, classId: classId, year: year) {
                teamRows.append(contentsOf: rows)
            }
        }

        let teamPages = await withBoundedConcurrency(teamRows, limit: Self.teamPageConcurrency) { row -> TeamPage? in
            guard let teamUrl = row.teamUrl else { return nil }
            return await self.fetchTeamPage(teamUrl: teamUrl, row: row, nowUtc: nowUtc)
        }

        var pageByRowIndex: [Int: TeamPage] = [:]
        for (index, page) in teamPages.enumerated() where page != nil {
            pageByRowIndex[index] = page
        }

        let standings: [StandingDraft] = teamRows.enumerated().map { index, row in
            let logoUrl = pageByRowIndex[index]?.logoUrl
            return StandingDraft(
                category: category,
                standingsClass: row.standingsClass,
                type: .team,
                entrantKey: "\(category.rawValue)-\(row.standingsClass.rawValue)-TEAM-\(row.teamName)-\(row.carNumber)",
                position: row.position,
                name: "#\(row.carNumber)",
                team: row.teamName,
                points: row.points,
                photoUrl: logoUrl,
                updatedAtUtc: nowUtc
            )
        }

        let carDrivers = pageByRowIndex.values.flatMap(\.drivers)

        return ImsaFetchResult(standings: standings, carDrivers: carDrivers)
    }

    /// Resuelve el classId de cada shortcode contra la API pública de clases — mapa
    /// vacío (nunca lanza) si esa API fallara, para caer al fallback.
    private func resolveClassIds(year: Int) async -> [String: String] {
        let url = "https://dvw6yynr86g3k.cloudfront.net/galaxy/api/classes?series_id=\(Self.seriesId)&season_id=\(year)"
        guard let html = try? await HTTPClient.fetchHTML(url) else { return [:] }
        return (try? parseClassIds(html)) ?? [:]
    }

    // 04/09/2026 (Fase 1 del diagnóstico): separado de resolveClassIds() para poder
    // testear el parsing contra un fixture JSON real sin red — ver
    // ImsaStandingsSourceTests.
    func parseClassIds(_ json: String) throws -> [String: String] {
        let classes = try JSONDecoder().decode([ImsaGalaxyClass].self, from: Data(json.utf8))
        var result: [String: String] = [:]
        for entry in classes {
            guard let shortcode = entry.shortcode else { continue }
            result[shortcode.uppercased()] = String(entry.id)
        }
        return result
    }

    private func fetchClassTeamRows(standingsClass: StandingsClass, classId: String, year: Int) async throws -> [TeamRow] {
        let html = try await HTTPClient.postForm(
            Self.ajaxUrl,
            fields: [
                "action": "getImsaStandings",
                "standings": "team",
                "seriesId": Self.seriesId,
                "currentYear": String(year),
                "classId": classId
            ],
            referer: "https://www.imsa.com/weathertech/standings/"
        )

        let doc = try SwiftSoup.parse(html)
        // Solo la tabla "desktop" trae puntos (td.totalpoints) — la "mobile" repite las
        // mismas filas sin esa columna.
        let rows = try doc.select("tr:has(td.totalpoints)").array()
        return try rows.enumerated().compactMap { index, row in
            try parseTeamRow(row, standingsClass: standingsClass, position: index + 1)
        }
    }

    /// "#04 Crowdstrike Racing by APR" -> número de coche + nombre de equipo. Algunas
    /// filas no traen el `<a>` de enlace (equipo sin ficha propia) — se lee igual el
    /// texto de la celda, esa fila se queda sin logo ni pilotos.
    // internal (no private): expuesta a test — ver ImsaStandingsSourceTests. TeamRow pasa
    // a internal por el mismo motivo.
    func parseTeamRow(_ row: Element, standingsClass: StandingsClass, position: Int) throws -> TeamRow? {
        guard let cell = try row.select("td.team-col").first() else { return nil }
        let text = try cell.text().trimmingCharacters(in: .whitespacesAndNewlines)
        guard let match = text.range(of: "^#(\\d+)\\s+(.+)$", options: .regularExpression) else { return nil }
        let matched = String(text[match])
        guard let carNumber = Self.firstGroup(of: "^#(\\d+)", in: matched),
              let teamName = Self.firstGroup(of: "^#\\d+\\s+(.+)$", in: matched)?.trimmingCharacters(in: .whitespacesAndNewlines),
              !teamName.isEmpty else {
            return nil
        }

        let points = try row.select("td.totalpoints").first().flatMap { try? $0.text().trimmingCharacters(in: .whitespacesAndNewlines) }
            .flatMap(Double.init) ?? 0.0
        let teamUrl = try cell.select("a.team-name").first().flatMap { try? $0.absUrl("href") }.flatMap { $0.isEmpty ? nil : $0 }

        return TeamRow(standingsClass: standingsClass, position: position, carNumber: carNumber, teamName: teamName, points: points, teamUrl: teamUrl)
    }

    private func fetchTeamPage(teamUrl: String, row: TeamRow, nowUtc: Date) async -> TeamPage? {
        do {
            // El Referer evita algún 403 puntual — basta con que la petición diga venir
            // de la propia imsa.com.
            let html = try await HTTPClient.fetchHTML(teamUrl, referer: "https://www.imsa.com/weathertech/teams/")
            return try parseTeamPage(html, teamUrl: teamUrl, standingsClass: row.standingsClass, carNumber: row.carNumber, nowUtc: nowUtc)
        } catch {
            return nil
        }
    }

    // 04/09/2026 (Fase 1 del diagnóstico): separado de fetchTeamPage() para poder testear
    // la detección de logo (descarta el logo fijo de la serie y el placeholder "nologo",
    // ninguno de los dos por posición fija) contra un fixture HTML sin red — ver
    // ImsaStandingsSourceTests.
    func parseTeamPage(_ html: String, teamUrl: String, standingsClass: StandingsClass, carNumber: String, nowUtc: Date) throws -> TeamPage {
        let doc = try SwiftSoup.parse(html, teamUrl)

        let logoCandidates = try doc.select(".team-logos img").array().compactMap { try? $0.absUrl("src") }
        let logoUrl = logoCandidates.first {
            !$0.localizedCaseInsensitiveContains("weathertech_championship") && !$0.localizedCaseInsensitiveContains("nologo")
        }

        let cards = try doc.select("div.imsa-card_item_widget").array()
        let drivers: [CarDriverDraft] = try cards.compactMap { card -> CarDriverDraft? in
            let name = try card.select("p.imsa-ciw-title").first()?.text().trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
            guard !name.isEmpty else { return nil }
            let photoUrl = try card.select("img.imsa-ciw-image").first().flatMap { try? $0.absUrl("data-src") }.flatMap { $0.isEmpty ? nil : $0 }

            return CarDriverDraft(
                category: .imsa,
                standingsClass: standingsClass,
                carNumber: carNumber,
                entryKey: TextNormalizer.normalize(name),
                name: name,
                photoUrl: photoUrl,
                updatedAtUtc: nowUtc
            )
        }

        return TeamPage(logoUrl: logoUrl, drivers: drivers)
    }

    private static func firstGroup(of pattern: String, in text: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        guard let match = regex.firstMatch(in: text, range: range), match.numberOfRanges > 1,
              let group = Range(match.range(at: 1), in: text) else {
            return nil
        }
        return String(text[group])
    }

    // internal (no private): expuestas a test — ver parseTeamRow/parseTeamPage arriba.
    struct TeamRow: Sendable {
        var standingsClass: StandingsClass
        var position: Int
        var carNumber: String
        var teamName: String
        var points: Double
        var teamUrl: String?
    }

    struct TeamPage: Sendable {
        var logoUrl: String?
        var drivers: [CarDriverDraft]
    }

    private static let ajaxUrl = "https://www.imsa.com/wp-admin/admin-ajax.php"
    /// WeatherTech SportsCar Championship (id 1; 2 y 3 son las otras dos series de IMSA).
    private static let seriesId = "1"
    /// No más de 8 páginas de equipo a la vez (de las ~45-50 en total).
    private static let teamPageConcurrency = 8

    private static let classShortcodes: [(StandingsClass, String)] = [
        (.gtp, "GTP"),
        (.lmp2, "LMP2"),
        (.gtdPro, "GTD PRO"),
        (.gtd, "GTD")
    ]

    /// Ids comprobados a mano para la temporada 2026 — solo se usan si la API de clases
    /// no responde.
    private static let fallbackClassIds: [StandingsClass: String] = [
        .gtp: "194",
        .lmp2: "196",
        .gtdPro: "192",
        .gtd: "191"
    ]
}

public struct ImsaFetchResult: Sendable {
    public var standings: [StandingDraft]
    public var carDrivers: [CarDriverDraft]
}

struct ImsaGalaxyClass: Decodable, Sendable {
    let id: Int
    let shortcode: String?
}
